impeller/renderer/backend/vulkan/sampler_library_vk.cc

```diff
@@ -16,15 +16,14 @@ SamplerLibraryVK::SamplerLibraryVK(
 
 SamplerLibraryVK::~SamplerLibraryVK() = default;
 
-std::shared_ptr<const Sampler> SamplerLibraryVK::GetSampler(
-    SamplerDescriptor desc) {
+const Sampler& SamplerLibraryVK::GetSampler(SamplerDescriptor desc) {
   auto found = samplers_.find(desc);
   if (found != samplers_.end()) {
-    return found->second;
+    return *found->second.get();
```

Comment:
Right here, this is a no-no.  You are transforming a `std::unique_ptr` to a reference.  References are non-null, std::unique_ptr are nullable.  If you return the reference to the unique_ptr you avoid that class of bugs.

Reply:
We cannot return a reference to a unique_ptr though?

Reply:
(from a hash map, specifically)

Reply:
Yea, you can.  Depending on the method you might have to do that trick where you have a static nullptr std::unique_ptr that you return the reference to sometimes.

Reply:
Here's an example:
```c++
#include <map>
#include <memory>

class Foo {};

static const std::unique_ptr<Foo> kNullFoo;

const std::unique_ptr<Foo>& find(const std::map<int, std::unique_ptr<Foo>>& map) {
  const auto& it = map.find(123);
  if (it == map.end()) {
    return kNullFoo;
  } else {
    return it->second;
  }
}
int main() {
  std::map<int, std::unique_ptr<Foo>> map;

  map.insert(std::make_pair(123, std::make_unique<Foo>()));

  const std::unique_ptr<Foo>& value = find(map);

  return 0;
}
```

Reply:
Using the types in sampler_library:

```c++
static const std::unique_ptr<Sampler>& find_thing(const SamplerMap& map, const SamplerDescriptor& desc) {
  const auto& it = map.find(desc);
  if (it == map.end()) {
    return kNullSampler;
  } else {
    return it->second;
  }
}
```

Gives

```
std::unique_ptr<const impeller::Sampler>' to 'unique_ptr<Sampler> &&' for 1st argument
unique_ptr.h(210, 3): Candidate template ignored: requirement 'is_convertible<const impeller::Sampler *, impeller::Sampler *>::value' was not satisfied [with _Up = const impeller::Sampler, _Ep = std::default_delete<const impeller::Sampler>]
unique_ptr.h(178, 12): Explicit constructor is not a candidate
```

Reply:
It looks like maybe you are trying to store references to unique_ptrs in the map?  Thats how you are getting the `&&`?  Can I see the whole patch or branch?

Reply:
Yeah one sec!

Reply:
Pushed it up

Reply:
You have to make it return a `const std::unique_ptr<const Sampler>&` to fix what you have right now.  I'm not sure if a `const Sampler` will work for you.  I don't know why that is off the bat

```diff
--- a/impeller/renderer/backend/vulkan/sampler_library_vk.cc
+++ b/impeller/renderer/backend/vulkan/sampler_library_vk.cc
@@ -16,9 +16,9 @@ SamplerLibraryVK::SamplerLibraryVK(
 
 SamplerLibraryVK::~SamplerLibraryVK() = default;
 
-static const std::unique_ptr<Sampler> kNullSampler = nullptr;
+static const std::unique_ptr<const Sampler> kNullSampler = nullptr;
 
-static const std::unique_ptr<Sampler>& find_thing(
+static const std::unique_ptr<const Sampler>& find_thing(
     const SamplerMap& map,
     const SamplerDescriptor& desc) {
   const auto& it = map.find(desc);
```

Reply:
(since you were using a `const Sampler&` it should be good)

Reply:
OH DUH

Reply:
OKAY! I DID IT

Thanks @gaaclarke I would have never figured that out!